Automattic/jetpack
# Add Mapkit to Maps block

## Proposed changes:

This PR brings support to use Mapkit as a map provider in the Map block. This has been quite an endeavor, and this PR served as the feature branch, where all PR's listed below were merged. Now it needs a final review, before being released.


| PR | URL | Status |
|---|------|--------|
| Add util functions to convert from/between Mapbox zoom levels & Mapkit camera distance | https://github.com/Automattic/jetpack/pull/29092 | 🟢 |
| Endpoint for Mapkit JWT signing | D101694-code | 🟢 | 
| Add Mapkit secret key constants | D103086-code | 🟢 | 
| Maps block: render Apple Maps maps on WP.com frontend | https://github.com/Automattic/jetpack/pull/29089 | 🟢 |
| Add getMapProvider function | https://github.com/Automattic/jetpack/pull/29116 | 🟢 |
| Add support for Mapkit geocoding | https://github.com/Automattic/jetpack/pull/29120 | 🟢 |
| Introduce useMapkitSetup hook | https://github.com/Automattic/jetpack/pull/29144 | 🟢 | 
| Add Mapkit support in the editor | https://github.com/Automattic/jetpack/pull/29252 | 🟢  |
| Only load Mapkit library block once per page | https://github.com/Automattic/jetpack/pull/29301 | 🟢 |
| Don't show Mapbox token input when map provider is Mapkit | https://github.com/Automattic/jetpack/pull/29306 | 🟢 |
| Only show specific controls when mapprovider is mapbox | https://github.com/Automattic/jetpack/pull/29333 | 🟢 |
| Round zoom value in block properties | https://github.com/Automattic/jetpack/pull/29334 |  🟢 |
| Recenter map after closing a marker info window | https://github.com/Automattic/jetpack/pull/29338 | 🟢 |
| Catch Mapkit API request failures & show error | https://github.com/Automattic/jetpack/pull/29356 | 🟢 |
| Move logic to getMapProvider | https://github.com/Automattic/jetpack/pull/29382 | 🟢 |
| Geocode address parameters in Mapkit | https://github.com/Automattic/jetpack/pull/29394 | 🟢 |
| Default to the default mapCenter defined in settings.js| https://github.com/Automattic/jetpack/pull/29385 | 🟢   |

![CleanShot 2023-04-20 at 12 56 19@2x](https://user-images.githubusercontent.com/528287/233346217-f31d2e69-f373-4c05-abde-079bb87dd116.png)

### Mapping provider

We'll only use Mapkit (Apple Maps) as a map provider on WPCOM Simple & Atomic sites. To determine which map provider is used, these steps are taken:

1. If the map style is terrain: use Mapbox
2. If a cookie called `map_provider` is set: use the value in the cookie (`mapbox` | `mapkit`)
3. If we still don't have a map provider:
    - On the post/page view: call `apply_filters( 'wpcom_map_block_map_provider', 'mapbox' );`. This will return whatever the filter returns on WPCOM, and fallback to the provided value mapbox on self-hosted JP sites.
    - In the editor: read the value of `window.map_block_map_provider`. This global variable is injected by [ETK](https://github.com/Automattic/wp-calypso/pull/75968). If it isn't present, fallback to Mapbox.

This approach will allow us to have the code already on WPCOM Simple & Atomic, and flip the switch later (by makeing a diff to WPCOM and a PR WPCOMSH).


### Functionality

The Mapkit integration offers the same functionality as Mapbox, except for:

- There's no option to show the map full-screen.
- There's no option to hide the street names.
- The `Black & White `style maps to Mapkit's `MutedStandard` style, which has some color.
- Mapkit doesn't offer the `Terrain` style, so when selecting that Mapbox will be used.

Other subtle differences:

- Mapkit's zoom & Mapbox's zoom levels don't map 100%. But it's pretty close.
- Mapkit uses its own geocoder when searching for addresses.

### Testing instructions

As can be seen in the overview above, (almost) all the changes have already been tested by Team Apex (and some by the Jetpack team) before being merged to this feature branch. 

The more eyes the better, so feel free to add a Map, add some points, see if everything behaves like it should, both when using Mapkit & Mapbox.

The easiest way to test is:

1. Run `bin/jetpack-downloader test jetpack add/map-block-mapkit` on your sandbox
2. Add this to `0-sandbox.php`:

```php
function wpcom_map_block_map_provider_mapkit() {
	return 'mapkit';
}

add_filter( 'wpcom_map_block_map_provider', 'wpcom_map_block_map_provider_mapkit', 20, 0 );
```
3. Thoroughly test the map block

Alternatively, you can test this on a self-hosted or JT site, by setting the cookie:

`document.cookie = 'map_provider=mapkit'`

When testing this on a JT or self-hosted site, make sure it's connected to Jetpack, otherwise the blog_id will be empty, and the call to fetch the API key will not work.

### Other information:


## Jetpack product discussion
pdDOJh-1tR-p2
## Does this pull request change what data or activity we track or use?
No

Fixes https://github.com/Automattic/wp-calypso/issues/73645, fixes https://github.com/Automattic/wp-calypso/issues/73641
